omegaconf/basecontainer.py

```diff
@@ -391,16 +397,18 @@ def _set_item_impl(self, key: Any, value: Any) -> None:
 
         elif isinstance(self.__dict__["_content"], list):
             target_node = isinstance(target_node_ref, ValueNode)
+        # We use set_value if:
+        # 1. Target node is a container and the value is MISSING or None
+        # 2. Target node is a container and has an explicit ref_type
+        # 3. Target node is a valuenode. If it's an AnyNode then if the
+        #    value is a primitive type it should set.
```

Comment:
These are not change requests:

for 1: reading this I am wondering why we don't always use set_value if the is_missing() or is_none()

for 2: it only makes sense to me that we use set_value on a container if the incoming value is a compatible. (dict, missing or None)

for 3: I am having a hard time parsing this one.

Reply:
> for 3: I am having a hard time parsing this one.

If the target  is a NodeValue then it should set his value. Furthermore if it's an AnyNode it should wrap when the input is a container and set when the input is an compatible type(primitive type).

Reply:
much better.